source/geod24/concurrency.d

```diff
@@ -926,84 +1038,138 @@ protected:
         ThreadInfo info;
 
         /// Semaphore reference that this Fiber is blocked on
-        FiberBinarySemaphore sem;
+        FiberBlocker blocker;
+
+        /// List of Resources held by this Fiber
+        DList!Resource resources;
 
-        this(void delegate() op, size_t sz = 512 * 1024) nothrow
+        this (void delegate() op, size_t sz = 512 * 1024) nothrow
         {
             super(op, sz);
         }
     }
 
-    protected class FiberBinarySemaphore
+    final public class FiberBlocker
     {
 
         /***********************************************************************
 
-            Associate `FiberBinarySemaphore` with the running `Fiber`
+            Associate `FiberBlocker` with the running `Fiber`
 
             `FiberScheduler` will check to see if the `Fiber` is blocking on a
-            `FiberBinarySemaphore` to avoid rescheduling it unnecessarily
+            `FiberBlocker` to avoid rescheduling it unnecessarily
 
         ***********************************************************************/
 
-        private void registerToInfoFiber(InfoFiber info_fiber = cast(InfoFiber) Fiber.getThis()) nothrow
+        private void registerToInfoFiber (InfoFiber info_fiber = cast(InfoFiber) Fiber.getThis()) nothrow
         {
             assert(info_fiber !is null, "This Fiber does not belong to FiberScheduler");
-            assert(info_fiber.sem is null, "This Fiber already has a registered FiberBinarySemaphore");
-            info_fiber.sem = this;
+            assert(info_fiber.blocker is null, "This Fiber already has a registered FiberBlocker");
+            info_fiber.blocker = this;
 
         }
 
-        void wait() nothrow
-        {
-            this.registerToInfoFiber();
-            FiberScheduler.yield();
-            scope (exit) this.notified = false;
-        }
+        /***********************************************************************
+
+            Wait on the blocker with optional timeout
+
+            Params:
+                period = Timeout period
 
-        bool wait(Duration period) nothrow
+        ***********************************************************************/
+
+        bool wait (Duration period = Duration.init) nothrow
         {
-            this.registerToInfoFiber();
-            this.limit = MonoTime.currTime + period;
-            FiberScheduler.yield();
+            if (period != Duration.init)
```

Comment:
What about `Duration.max`?

Reply:
I mean, I think I saw code which handled both cases.

Reply:
Only `Duration.init` is considered infinite timeout throughout localrest and concurrency.d. You must have seen it elsewhere?